add docs for de-supporting an os (#219)
mozilla-releng/relengdocs: balrog/os-de-support.rst

Typically NO-UPDATE is used to stop updates temporarily just before the first unsupported version is available on each channel. Once the last supported version is available, an ESR-SWITCH or DE-SUPPORT rule is implemented and the NO-UPDATE rule is deleted.

Update Balrog rules for Nightly
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For the Nightly channel, we usually DE-SUPPORT.

Just *before* the first nightly build (could be a few days earlier, if convenient) of the first unsupported version, pause Nightly updates for the affected OS version: 

* Create a new rule matching the ``Firefox : Nightly*`` channel and the OS version to serve the ``NoUpdate`` release.

Shortly *after* the first nightly build of the first unsupported version:

* Determine the build ID of the last build of the last supported version: You might check `archive.mozilla.org <https://archive.mozilla.org/pub/firefox/nightly/>`__ for instance.
* Create a rule to pin Nightly* users to the most recent supported version: match the ``Firefox : Nightly*`` channel, the affected OS version and build ID < latest supported build ID determined above; serve the Release corresponding to that build ID.

> * Update the Release referenced by the new pinning rule, to include aliases (this is only required on Nightly, as it is caused by `bug 1810740 <https://bugzilla.mozilla.org/show_bug.cgi?id=1810740>`__): Update the Release blob by adding aliases like::

Reviewer: Maybe move this above the rule creation so we don't have a window where the rule points at the partial release blob.
Also, we might want to mention that the blob is called `Firefox-mozilla-central-nightly-<buildid>`

> * Update the Release referenced by the new pinning rule, to include aliases (this is only required on Nightly, as it is caused by `bug 1810740 <https://bugzilla.mozilla.org/show_bug.cgi?id=1810740>`__): Update the Release blob by adding aliases like::

Reviewer: Let's also explicitly says the up-to-date list of aliases should be looked up in Firefox-mozilla-central-nightly-latest, just in case they change before we have to do this again.

    "Darwin_aarch64-gcc3": {
      "alias": "Darwin_x86_64-gcc3-u-i386-x86_64"
    },
    "Darwin_x86-gcc3": {
      "alias": "Darwin_x86_64-gcc3-u-i386-x86_64"
    },
    "Darwin_x86-gcc3-u-i386-x86_64": {
      "alias": "Darwin_x86_64-gcc3-u-i386-x86_64"
    },
    "Darwin_x86_64-gcc3": {
      "alias": "Darwin_x86_64-gcc3-u-i386-x86_64"
    },
    "WINNT_x86-msvc-x64": {
      "alias": "WINNT_x86-msvc"
    },
    "WINNT_x86-msvc-x86": {
      "alias": "WINNT_x86-msvc"
    },    
    "WINNT_x86_64-msvc-x64": {
      "alias": "WINNT_x86_64-msvc"
    },    

* Create a new De-Support release referencing the SUMO KB article. Careful: Use a locale-agnostic link like https://support.mozilla.org/kb/firefox-users-macos-1012-1013-1014-moving-to-extended-support (instead of https://support.mozilla.org/en-US/kb/firefox-users-macos-1012-1013-1014-moving-to-extended-support). Example release blob: ::

    {
        "detailsUrl": "https://support.mozilla.org/kb/firefox-users-macos-1012-1013-1014-moving-to-extended-support",
        "displayVersion": "115.0",
        "name": "OSX-10.12-10.14-Desupport",
        "product": "Firefox",
        "schema_version": 50
    }

> * Create a rule to serve the de-support notice: match the ``Firefox : Nightly*`` channel and the affected OS version with lower priority than the pinning rule (so probably those who have been updated to the last supported build); serve the new De-Support release.

Reviewer: ```suggestion
* Create a rule to serve the de-support notice: match the ``Firefox : nightly*`` channel and the affected OS version with lower priority than the pinning rule (so probably those who have been updated to the last supported build); serve the new De-Support release.
```

> * Delete the ``NoUpdate`` rule created earlier.

Reviewer: ```suggestion
* Delete the ``No-Update`` rule created earlier.
```

Update Balrog rules for Beta and Aurora
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Check with the *Product* organization to verify per-channel requirements for Beta and DevEdition.

To implement DE-SUPPORT for DevEdition, follow the de-support procedure for Nightly, above, with these changes:

* Make changes just before and after the Merge Day II merge of central to beta.
* Use the ``Firefox : aurora*`` channel
* There should be no need to add aliases to the release blob.

To implement ESR-SWITCH for Beta, follow the de-support procedure for Release, below, with these changes:

* Make changes just before and after the Merge Day II merge of central to beta.
* Use the ``Firefox : beta*`` channel
* There should be no need to add aliases to the release blob.

Update Balrog rules for Release
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For the Release channel, we generally ESR-SWITCH: serve an update referencing the channel-switching MAR, so that
users on the affected OS move from the Release channel to the ESR channel.

> Just *before* the first release build of the first unsupported version, pause Release updates for the affected OS version. This will typically be just before the Merge Day I merge of Beta to Release, for the release after the Beta changes.

Reviewer: Probably want to change this section so that:
- ahead of time, create the esr-switch release blob, with a pointer to your script
- at the start of RC week, set up the ESR-SWITCH rule on `beta-localtest` and `release-localtest`
- we get QA to verify the migration
- on merge day II, set it up on `beta-cdntest` and `release-cdntest`
- on release day, set it up for `beta*` and `release*`
?

> * Create a new rule matching the ``Firefox : Release*`` channel and the OS version to serve the ``NoUpdate`` release.

Reviewer: ```suggestion
* Create a new rule matching the ``Firefox : release*`` channel and the OS version to serve the ``No-Update`` release.
```

> * Create a new rule matching the ``Firefox : Release*`` channel and the OS version to serve the ``NoUpdate`` release.

Reviewer: I don't think we need to go through No-Update here, just set up the channel-switch rule at release time?

Shortly *after* the first release build of the first unsupported version:

* Create a new release to serve the channel-switching mar. *TBD: do we duplicate an existing release, then manually replace the url?*
* Create a new rule matching the ``Firefox : Release*`` channel and the OS version to serve the channel-switching release.
* *TBD: Is that all, or do we need pinning?*
* Delete the ``NoUpdate`` rule created earlier.
 
Update Balrog rules for ESR
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> TBD

Reviewer: This should be similar to nightly and aurora, pin to latest supported release and de-support.

* Like https://bugzilla.mozilla.org/show_bug.cgi?id=1275609 ?
* Pin and de-support?

> * How do we remember to do this? File a bug now and .... ?

Reviewer: When an esr branch goes EOL we should go through the release channel rules between that esr major version and the next, to figure out whether we need watersheds etc, and the no-longer-supported OS versions should come up at that point.

Author: Thanks, that makes sense. In that case, it sounds like that process will usually just work, and I don't see any point in describing it here. I'll keep this section brief, at least for now.
